### PR TITLE
Granting wish part of #15497: link the source file loaded by "Load" in coqdoc when this file is referred by qualid

### LINKED ISSUE #15497: .glob files are incorrect (and result in incorrect html linking) when `Load` is used
#### Description of the problem

Consider the following files:
`A.v`:
```coq
Require Import Init.
Require Import Init.
```
`B.v`:
```coq
Require Import Init. Local Open Scope nat_scope.
Require Import Init.
Load A.
Check nat.
```
`C.v`:
```coq
Require Import Init. Local Open Scope nat_scope.
Require Import Init.
(*****)
Check nat.
```
If I run `coq_makefile -Q . Top A.v B.v C.v -o Makefile` and then `make clean` and then `make all html`, I get the following `.glob` files:
`A.glob`:
```
DIGEST d59f3182ed63bdbcf24de6c7917f9699
FTop.A
R15:18 Coq.Classes.Init <> <> lib
R36:39 Coq.Classes.Init <> <> lib
```
`B.glob`:
```
DIGEST 4f20981d9fdb7f179fb549820cab8dea
FTop.B
R15:18 Coq.Classes.Init <> <> lib
R64:67 Coq.Classes.Init <> <> lib
R15:18 Coq.Classes.Init <> <> lib
R36:39 Coq.Classes.Init <> <> lib
R84:86 Coq.Init.Datatypes <> nat ind
```
`C.glob`:
```
DIGEST 5d05a02f5187592c07d44a384272f7f8
FTop.C
R15:18 Coq.Classes.Init <> <> lib
R64:67 Coq.Classes.Init <> <> lib
R84:86 Coq.Init.Datatypes <> nat ind
```

Note that `B.glob` should be the same as `C.glob`, but instead it has all of the references from `A.v`.  This means that the generated html contains incorrect links:
<details><summary>Top.B.html source</summary>

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link href="coqdoc.css" rel="stylesheet" type="text/css" />
<title>Top.B</title>
</head>

<body>

<div id="page">

<div id="header">
</div>

<div id="main">

<h1 class="libtitle">Library Top.B</h1>

<div class="code">
<span class="id" title="keyword">Require</span> <span class="id" title="keyword">Import</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">Init</span></a>. <span class="id" title="var">Local Open</span> <span class="id" title="keyword">Scope</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">nat_scope</span></a>.<br/>
<span class="id" title="keyword">Require</span> <span class="id" title="keyword">Import</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">Init</span></a>.<br/>
<span class="id" title="keyword">Load</span> <span class="id" title="var">A</span>.<br/>
<span class="id" title="keyword">Check</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Init.Datatypes.html#nat"><span class="id" title="inductive">nat</span></a>.<br/>
</div>
</div>

<div id="footer">
<hr/><a href="index.html">Index</a><hr/>This page has been generated by <a href="http://coq.inria.fr/">coqdoc</a>
</div>

</div>

</body>
</html>
```
</details>

<details><summary>Top.C.html source</summary>

```html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link href="coqdoc.css" rel="stylesheet" type="text/css" />
<title>Top.C</title>
</head>

<body>

<div id="page">

<div id="header">
</div>

<div id="main">

<h1 class="libtitle">Library Top.C</h1>

<div class="code">
<span class="id" title="keyword">Require</span> <span class="id" title="keyword">Import</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">Init</span></a>. <span class="id" title="var">Local Open</span> <span class="id" title="keyword">Scope</span> <span class="id" title="var">nat_scope</span>.<br/>
<span class="id" title="keyword">Require</span> <span class="id" title="keyword">Import</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">Init</span></a>.<br/>
<span class="id" title="keyword">Check</span> <a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Init.Datatypes.html#nat"><span class="id" title="inductive">nat</span></a>.<br/>
</div>
</div>

<div id="footer">
<hr/><a href="index.html">Index</a><hr/>This page has been generated by <a href="http://coq.inria.fr/">coqdoc</a>
</div>

</div>

</body>
</html>
```
</details>

`Top.B.html`:
image
vs
`Top.C.html`:
image
Note that `nat_scope` should not be linked (as it is not in C), but in B it is linked as `<a class="idref" href="http://coq.inria.fr/distrib/V8.14.0/stdlib//Coq.Classes.Init.html#"><span class="id" title="library">nat_scope</span></a>`.

#### Coq Version

8.14

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

Depends on #15511 and #15512.

Closes #15497 (wish part)

For each `Load qualid.` line in a file, we add a line of the form `Rloc1:loc2 fullqualid <> <> lib` in the glob file so that it is hyperlinked by coqdoc.

- [x] Added / updated **test-suite**.
- [ ] Added **changelog**.

### CODE REVIEW

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 1]
```diff
@@ -479,19 +479,21 @@ toplevel. This kind of file is called a *script* for Coq. The standard
...
 
 
-.. cmd:: Load {? Verbose } {| @string | @ident }
+.. cmd:: Load {? Verbose } {| @string | @qualid }
 
-   Loads a file.  If :n:`@ident` is specified, the command loads a file
-   named :n:`@ident.v`, searching successively in
-   each of the directories specified in the *loadpath*. (see Section
-   :ref:`libraries-and-filesystem`)
+   Executes the commands of a file in the current context.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:484]:
```suggestion
   Executes the commands of a :n:`.v` file in the current context.
```

Is that right--only .v files can be loaded?  If not, is it really useful to load non-.v files?

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 2]
```diff
@@ -479,19 +479,21 @@ toplevel. This kind of file is called a *script* for Coq. The standard
...
+   If :n:`@qualid` is specified, :n:`Load` searches
+   in the *loadpath* (see Section :ref:`libraries-and-filesystem`)
+   for a file with suffix :n:`.v` whose logical name is :n:`@qualid`.
 
    If :n:`@string` is specified, it must specify a complete filename.
    `~` and .. abbreviations are
    allowed as well as shell variables. If no extension is specified, Coq
-   will use the default extension ``.v``.
+   will use the default extension ``.v``. If the filename is relative,
+   it is searched in the loadpath, ignoring the logical part.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:494]:
```suggestion
   appends the default extension ``.v``. If the filename is relative,
   Coq searches in the loadpath, ignoring the logical part of :n:`@qualid`.
```

"ignoring the logical part" -- You mean (for example) strip off `Coq.` from `Coq.Init.Nat`.  Can only be the first component, maybe can say that more explicitly?

File: doc/sphinx/proof-engine/vernacular-commands.rst [hunk 3]
```diff
@@ -501,7 +503,7 @@ toplevel. This kind of file is called a *script* for Coq. The standard
    :n:`Verbose` displays the Coq output for each command and tactic
    in the loaded file, as if the commands and tactics were entered interactively.
 
-   .. exn:: Can’t find file @ident on loadpath.
+   .. exn:: Can’t find file @qualid on loadpath.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/vernacular-commands.rst:506]: Maybe shouldn't refer to `@qualid` as a file?

**Review by jfehrle — COMMENTED** [review]
A couple suggestions

**Comment by JasonGross** [comment]
Can you also link the string when there's an equivalent qualid that would work (i.e., when the file is a .v file and exists in the library search path)?  This way the minimizer could automatically qualify the Load from a string to a qualid.

File: doc/sphinx/proof-engine/vernacular-commands.rst (hunk 1 above)
Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst:484]: The contents of the file must always be something that's valid for a .v file.  The ability to load files with different extensions might be useful?

Comment (JasonGross) [reply doc/sphinx/proof-engine/vernacular-commands.rst:484]: In particular, if you can Load .out files, then you can combine `Redirect` with `Load` to dynamically emit Gallina (e.g., via Ltac `idtac`) and then run it.

Comment (jfehrle) [reply doc/sphinx/proof-engine/vernacular-commands.rst:484]: Gotta be careful about "might be useful".
FWIW, in coqtop, Redirect silently overwrites any existing file.  It doesn't do an append.
In CoqIDE, Redirect creates an empty file and prints the idtac in the Messages panel.

**Comment by JasonGross** [comment]
@herbelin what was this blocking on?

**Comment by SkySkimmer** [comment]
#15512

**Comment by ppedrot** [comment]
I still advocate for killing Load, not extending it, and I believe I am not the only one around.

**Pull request closed without merging**